tests/pyfunc/test_spark.py

```diff
@@ -1298,3 +1298,59 @@ def predict(self, context, model_input, params=None):
 
         res = spark_df.withColumn("res", udf("input_col")).select("res").toPandas()
         assert res["res"][0] == ("string")
+
+
+def test_spark_udf_structs_and_arrays(spark, tmp_path):
+    class MyModel(mlflow.pyfunc.PythonModel):
+        def predict(self, context, model_input):
+            return [str(" | ".join(map(str, row))) for _, row in model_input.iterrows()]
+
+    df = spark.createDataFrame(
+        [
+            (
+                "a",
+                [0],
+                {"bool": True},
+                [{"double": 0.1}],
+            )
+        ],
+        schema=StructType(
+            [
+                StructField(
+                    "str",
+                    StringType(),
+                ),
+                StructField(
+                    "arr",
+                    ArrayType(IntegerType()),
+                ),
+                StructField(
+                    "obj",
+                    StructType(
+                        [
+                            StructField("bool", BooleanType()),
+                        ]
+                    ),
+                ),
+                StructField(
+                    "obj_arr",
+                    ArrayType(
+                        StructType(
+                            [
+                                StructField("double", DoubleType()),
+                            ]
+                        )
+                    ),
+                ),
+            ]
+        ),
+    )
+    mlflow.pyfunc.save_model(
+        path=tmp_path,
+        python_model=MyModel(),
+        signature=mlflow.models.infer_signature(df),
+    )
+
+    udf = mlflow.pyfunc.spark_udf(spark=spark, model_uri=tmp_path, result_type="string")
+    pdf = df.withColumn("output", udf("str", "arr", "obj", "obj_arr")).toPandas()
+    assert pdf["output"][0] == "a | [0] | {'bool': True} | [{'double': 0.1}]"
```

Comment:
This looks good! Can we add more tests for more complex structures and more rows?

Reply:
Adding more tests

Reply:
Done!